### ISSUE TITLE
New implementation of ls()

### ISSUE THREAD

**Issue by frankenjoe** [audeering/audbackend#40]
In #39 we discussed that the current implementation does not make sense as it returns paths in backend format.  Here is a first attempt how an alternative implementation could look like. It introduces a new function `Backend.find()` (for simplicity it's implemented in the child class here). We need a new name as it has a different signature than `ls()`. As mentioned in #39 it also introduces a new abstract method `_inverse_path()` to convert the matched files to the virtual file format:

```python
class FileSystem(audbackend.FileSystem):

    def _inverse_path(
            self,
            path: str,  # backend path
    ) -> (str, str, str):
        r"""Convert backend path to virtual file path."""
        tokens = path.split(os.path.sep)
        folder = os.path.sep.join(tokens[:-3])
        name = tokens[-3]
        version = tokens[-2]
        ext = tokens[-1][len(name) + len(version) + 1:]
        file = f'{name}{ext}'
        return folder, file, version

    def find(
            self,
            *,
            version: str = None,
            folder: str = None,
            pattern: str = None,
    ):
        r"""Will be implemented in Backend"""
        folder = folder or ''
        ls = self._find(folder)
        match = set()
        for path in ls:
            f, n, v = self._inverse_path(path)
            if version is not None and v != version:
                continue
            if pattern is not None and not fnmatch.fnmatch(n, pattern):
                continue
            match.add(self.join(f, n))
        return sorted(list(match))

    def _find(
            self,
            folder: str,  # virtual path
    ):
        r"""List content of folder."""
        folder = folder.replace(  # convert to backend path
            self.sep,
            os.path.sep,
        )
        root = audeer.path(
            self.host,
            self.repository,
            folder,
        )
        match = audeer.list_file_names(
            root,
            basenames=True,
            recursive=True,
        )
        return [os.path.join(folder, x) for x in match]
```

Here is a test:

```python
FileSystem('.', 'fs')
backend.put_file('test.txt', 'sub/dir/test', '1.0.0')
backend.put_file('test.txt', 'test.txt', '1.0.0')
backend.put_file('test.txt', 'sub/dir/test', '1.1.0')

ret = backend.find()
print(ret)
ret = backend.find(folder='sub')
print(ret)
ret = backend.find(version='1.1.0')
print(ret)
ret = backend.find(pattern='*.txt')
print(ret)
```
```
['sub/dir/test', 'test.txt']
['sub/dir/test']
['sub/dir/test']
['test.txt']
```

**Comment by hagenw** [comment]
Great to see that it can work.

What I don't exactly get is the difference between `find()` and `ls()`. I understand that one is called with `version`, `folder`, `pattern`, whereas the other is called with `path`, but I think we can achieve exactly the same with both functions.
In the following code block I mention possible corresponding `ls()` calls.

```python
backend.find()
backend.ls('/', recursive=True)

backend.find(folder='sub')
backend.ls('/sub', recursive=True)

backend.find(version='1.1.0')
backend.ls('/', version='1.1.0', recursive=True)

backend.find(pattern='*.txt')
backend.ls('*.txt')
```

**Comment by frankenjoe** [comment]
> whereas the other is called with `path`, but I think we can achieve exactly the same with both functions.

But then the result will still be different.

**Comment by frankenjoe** [comment]
`ls()` currently returns in backend format and also only the content of a folder and not all files it finds below the folder.

**Comment by hagenw** [comment]
> `ls()` currently returns in backend format and also only the content of a folder and not all files it finds below the folder.

Yes, but I doubt that we can implement the current `ls()` for every backend, so I would say the current implementation is a bug and we need to change it anyway.
But I agree with `find()` it would be easier to simply mark `ls()` as deprecated and remove it at a later version.

**Comment by frankenjoe** [comment]
Solved with #53
